File: app/src/main/java/com/stathis/elmepaunivapp/ui/syllabus/SyllabusActivity.java
package com.stathis.elmepaunivapp.ui.syllabus;

import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.util.Log;

import com.google.android.material.tabs.TabLayout;
import com.google.gson.Gson;
import com.stathis.elmepaunivapp.R;
import com.stathis.elmepaunivapp.abstraction.AbstractActivity;
import com.stathis.elmepaunivapp.listeners.activity_listeners.SyllabusActivityListener;
import com.stathis.elmepaunivapp.ui.syllabus.model.Semester;
import com.stathis.elmepaunivapp.ui.syllabus_lessons.SyllabusLessonsActivity;

public class SyllabusActivity extends AbstractActivity implements SyllabusActivityListener {

    private TabLayout tabLayout;
    private RecyclerView recyclerView;
    private SyllabusViewModel viewModel;

    public SyllabusActivity() {
        super(R.layout.activity_syllabus);
    }

    @Override
    public void initial() {
        viewModel = new ViewModelProvider(this).get(SyllabusViewModel.class);

        recyclerView = findViewById(R.id.syllabus_recycler);
        tabLayout = findViewById(R.id.tabLayout);
    }

    @Override
    public void running() {
        int userChoice = getIntent().getIntExtra("userTabChoice", 0);

        viewModel.initListener(this);
        recyclerView.setAdapter(viewModel.adapter);

        tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                switch (tab.getPosition()) {
                    case 0: {
                        viewModel.getDataList(getApplicationContext());
                        viewModel.adapter.notifyDataSetChanged();
                        break;
                    }
                    case 1: {
                        viewModel.getBusinessAdministrationList(getApplicationContext());
                        viewModel.adapter.notifyDataSetChanged();
                        break;
                    }
                    case 2: {
                        viewModel.getMarketingList(getApplicationContext());
                        viewModel.adapter.notifyDataSetChanged();
                        break;
                    }
                }

            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {}

            @Override
            public void onTabReselected(TabLayout.Tab tab) {}
        });

        tabLayout.getTabAt(userChoice).select();
        viewModel.getDataList(getApplicationContext());
    }

    @Override
    public void stopped() {}

    @Override
    public void showLessons(Semester data) {
        String jsonArray = new Gson().toJson(data.getLessons());
        Log.d("JSON_ARRAY", jsonArray);
        startActivity(new Intent(this, SyllabusLessonsActivity.class)
                .putExtra("ARRAY", jsonArray)
                .putExtra("LESSONS_INFO", data.getLessonInfo()));
    }
}

File: app/src/main/java/com/stathis/elmepaunivapp/ui/syllabus/SyllabusActivity.kt
package com.stathis.elmepaunivapp.ui.syllabus

import android.content.Intent
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.tabs.TabLayout
import com.stathis.elmepaunivapp.R
import com.stathis.elmepaunivapp.abstraction.ElmepaBindingActivity
import com.stathis.elmepaunivapp.callbacks.SyllabusClickListener
import com.stathis.elmepaunivapp.databinding.ActivitySyllabusBinding
import com.stathis.elmepaunivapp.ui.syllabus.model.Semester
import com.stathis.elmepaunivapp.ui.syllabus_lessons.SyllabusLessonsActivity
import kotlinx.android.synthetic.main.activity_syllabus.*

class SyllabusActivity : ElmepaBindingActivity<ActivitySyllabusBinding>(R.layout.activity_syllabus) {

    private lateinit var viewModel: SyllabusViewModel

    override fun init() {
        viewModel = ViewModelProvider(this).get(SyllabusViewModel::class.java)
    }

    override fun startOps() {
        binding.syllabusRecycler.adapter = viewModel.adapter

        val userChoice = intent.getIntExtra(resources.getString(R.string.userTabChoice), 0)

        viewModel.observe(this)
        viewModel.error.observe(this, Observer{
            when(it){
                true -> Snackbar.make(binding.syllabusScreenParent,resources.getString(R.string.announcements_get_error), Snackbar.LENGTH_LONG).show()
                false -> Unit
            }
        })

        binding.tabLayout.getTabAt(userChoice)?.select()

        viewModel.bindCallback(object : SyllabusClickListener {
            override fun onSemesterTap(syllabus: Semester) {
                startActivity(Intent(this@SyllabusActivity, SyllabusLessonsActivity::class.java).putExtra(getString(R.string.syllabus_intent_data), syllabus))
            }
        })

        binding.tabLayout.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab?) {
                when (tab?.position) {
                    0 -> viewModel.getDataList()
                    1 -> viewModel.getBaList()
                    2 -> viewModel.getMktList()
                    else -> Unit
                }
            }

            override fun onTabUnselected(tab: TabLayout.Tab?) {}
            override fun onTabReselected(tab: TabLayout.Tab?) {}
        })
    }

    override fun stopOps() {}
}